cluster/images/etcd/migrate/integration_test.go

```diff
@@ -0,0 +1,348 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	cryptorand "crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"math/big"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+)
+
+var (
+	testSupportedVersions = MustParseSupportedVersions("2.2.1, 2.3.7, 3.0.17, 3.1.12")
+	testVersionOldest     = &EtcdVersion{2, 2, 1}
+	testVersionPrevious   = &EtcdVersion{3, 0, 17}
+	testVersionLatest     = &EtcdVersion{3, 1, 12}
+)
+
+func TestMigrate(t *testing.T) {
+	migrations := []struct {
+		title        string
+		memberCount  int
+		startVersion string
+		endVersion   string
+		protocol     string
+	}{
+		// upgrades
+		{"v2-v3-up", 1, "2.2.1/etcd2", "3.0.17/etcd3", "https"},
+		{"v3-v3-up", 1, "3.0.17/etcd3", "3.1.12/etcd3", "https"},
+		{"oldest-newest-up", 1, "2.2.1/etcd2", "3.1.12/etcd3", "https"},
+
+		// warning: v2->v3 ha upgrades not currently supported.
+		{"ha-v3-v3-up", 3, "3.0.17/etcd3", "3.1.12/etcd3", "https"},
+
+		// downgrades
+		{"v3-v2-down", 1, "3.0.17/etcd3", "2.2.1/etcd2", "https"},
+		{"v3-v3-down", 1, "3.1.12/etcd3", "3.0.17/etcd3", "https"},
+
+		// warning: ha downgrades not yet supported.
+	}
+
+	for _, m := range migrations {
+		t.Run(m.title, func(t *testing.T) {
+			start := MustParseEtcdVersionPair(m.startVersion)
+			end := MustParseEtcdVersionPair(m.endVersion)
+
+			testCfgs := clusterConfig(t, m.title, m.memberCount, m.protocol)
+
+			servers := []*EtcdMigrateServer{}
+			for _, cfg := range testCfgs {
+				client, err := NewEtcdMigrateClient(cfg)
+				if err != nil {
+					t.Fatalf("Failed to create client: %s", err)
+				}
+				server := NewEtcdMigrateServer(cfg, client)
+				servers = append(servers, server)
+			}
+
+			// Start the servers.
+			parallel(servers, func(server *EtcdMigrateServer) {
+				dataDir, err := OpenOrCreateDataDirectory(server.cfg.dataDirectory)
+				if err != nil {
+					t.Fatalf("Error opening or creating data directory %s: %s", server.cfg.dataDirectory, err)
+				}
+				migrator := &Migrator{server.cfg, dataDir, server.client}
+				err = migrator.MigrateIfNeeded(start)
+				if err != nil {
+					t.Fatalf("Migration failed: %s", err)
+				}
+				err = server.Start(start.version)
+				if err != nil {
+					t.Fatalf("Failed to start server: %s", err)
+				}
+			})
+
+			// Write a value to each server, read it back.
+			parallel(servers, func(server *EtcdMigrateServer) {
+				err := server.client.Put(start.version, "/registry/k1", "value1")
+				if err != nil {
+					t.Fatalf("failed to write text value: %s", err)
+				}
+
+				value, err := server.client.Get(start.version, "/registry/k1")
```

Comment:
If we do write different data on each server, would it make sense to break out the read check so we can check readability of all keys on each server?

Reply:
Yes, we'll check that all test values can be read back from all the servers. Thanks!